xarray/coding/cftimeindex.py

```diff
@@ -215,7 +223,8 @@ def get_date_type(self):
 
 
 def assert_all_valid_date_type(data):
-    import cftime
+    if cftime is None:
+        raise ModuleNotFoundError("No module named 'cftime'")
```

Comment:
I *think* this may be the only place where this import could be attempted regularly without `cftime` installed.  Would only making a change here fix the performance issue?

Reply:
I'd have thought its better to avoid a pattern with these perfomance characteristics in general - hence why I changed it for all instances, rather than just the one that was causing me issues. I think its pretty clearly a low risk change, so thought its more a question of what's a better pattern to follow going forward, rather than what currently falls in the critical path.

Reply:
I guess my thinking was, at least in the case of the `cftime_offsets.py` module, if it is a performance issue to attempt importing `cftime` when it is not installed, why attempt it -- even once -- if we know we will never need to?  Note that we follow this current pattern in more than just the `cftime_offsets.py` and `cftimeindex.py` modules.  There are places in `times.py` where we import `cftime` within functions, as well as numerous places in the tests (again though, these are places where I do not think we would see material impacts on performance).  I'm open to discussing this more, however.

One other place where I do think this new pattern could positively impact performance is linked below; this is another example of where we might attempt importing `cftime` regularly when it is not installed.  It would be great if you could modify the import logic there to be more performance-friendly too.
https://github.com/pydata/xarray/blob/35d798a36912418e5c6b4bd7d6c3c6e39accf3d0/xarray/plot/utils.py#L625-L630